### PR TITLE
Improve `outlines.processors`, add integration tests to test_generate.py

### PR DESCRIPTION
**Pull request by lapp0**
A lot of these fixes were intended for https://github.com/outlines-dev/outlines/pull/966 however that's blocked until there's a new `transformers` release. 

These improvements are general to all models and will enable PRs resolving https://github.com/outlines-dev/outlines/issues/806 and https://github.com/outlines-dev/outlines/issues/965

# Structure of `OutlinesLogitsProcessor`

The goal is to create a base class which allows a logits processors to be implemented once and used for any `outlines.models` inference library.

To accomplish this we must normalize the input array. It must have a consistent type (`torch.Tensor`) and consistent dimensionality (2). We can normalize both of these simply, and without any copy operations.

`mlx.core.array`, `numpy.array`, and `torch.Tensor` all support [pythons array standard `__dlpack__`](https://data-apis.org/array-api/latest/API_specification/generated/array_api.array.__dlpack__.html). This standard allows for casting between array types without copying.

`torch.Tensor` is the only input type which cannot always be cast to any other type because torch tensors may live in GPU memory. Therefore, we cast all arrays to `torch.Tensor`, implement logits processors using torch methods, and convert back to the original array type in `OutlinesLogitsProcessor`. See docstring of `OutlinesLogitsProcessor.__call__()` for more details.

# Detailed Changes
- Rename `BaseLogitsProcessor` to `OutlinesLogitsProcessor`
- Ensure `OutlinesLogitsProcessor.process_logits()` is always passed a 2D batch request with `torch.Tensor` logits and `List` input_ids. Also clean up code to be more readable in `OutlinesLogitsProcessor__call__()`
- Ensure `FSMLogitsProcessor` allows unstable sequence ordering (beam search in transformers and vLLM change the order of sequences)
- Update `tests/generate/test_generate.py` to cover more permutations of 
  - regex / text 
  - batch / single
  - greedy / multinomial / beam search
  - `stream()` / `generate()`
- Ensure performance stability with difference array libraries through `benchmark_processors.py`

### CODE REVIEW

File: outlines/processors/base_logits_processor.py [hunk 1]
```diff
@@ -14,10 +14,11 @@ def is_mlx_array(logits):
     return isinstance(logits, mx.array)
 
 
-class BaseLogitsProcessor(Protocol):
+class OutlinesLogitsProcessor(Protocol):
```
Comment (rlouf) [inline outlines/processors/base_logits_processor.py:17]: Why not call it `LogitsProcessor`?

File: outlines/processors/base_logits_processor.py [hunk 2]
```diff
@@ -29,50 +30,100 @@ class BaseLogitsProcessor(Protocol):
...
-                return mx.array(logits_float32_numpy)
-
-            else:
-                raise TypeError(
-                    "LogitsProcessor must be called with either np.NDArray"
-                    ", torch.Tensor, or mlx.core.array typed logits"
-                )
+
+        # ensure logits are torch Tensors
+        torch_logits = self._to_torch(logits)
```
Comment (rlouf) [inline outlines/processors/base_logits_processor.py:65]: What is the cost of this conversion? We need to profile this.

File: outlines/processors/base_logits_processor.py [hunk 3]
```diff
@@ -29,50 +30,100 @@ class BaseLogitsProcessor(Protocol):
...
+        - logits processors are only used once and never re-applied for a new sequence generator
+        - Some models only pass output_ids, some models such as llamacpp and transformers prefix with input_ids
+        - Some sampling methods, such as beam search, result in unstable sequence ordering in models like vLLM
+        """
+        pass
 
+    @torch.no_grad()
     def __call__(
         self,
         input_ids: Union[NDArray[np.int64], List[int], torch.Tensor],
```
Comment (rlouf) [inline outlines/processors/base_logits_processor.py:51]: It's probably time to create a type alias that include the MLX tensor type.

File: outlines/processors/structured.py [hunk 4]
```diff
@@ -63,12 +63,13 @@ def __init__(self, tokenizer: "Tokenizer", fsm: Guide):
             The finite state machine which is used to bias the logits.
         """
         self.tokenizer = tokenizer
-        self._fsm_state = 0
+        self._fsm_states: Dict[int, int] = {}
         self.fsm: Guide = fsm
         self._is_first_token = True
+        self._seq_start_idx: Optional[int] = None
 
     def process_logits(
```
Comment (rlouf) [inline outlines/processors/structured.py:71]: cf above remark on dispatching the function depending on the type of the tensors.

File: outlines/processors/base_logits_processor.py (hunk 1 above)
Comment (rlouf) [reply outlines/processors/base_logits_processor.py:17]: Actually, do we really need that class if all logit processors are ever going to be `FSMLogitsProcessors`

**Review by rlouf — CHANGES_REQUESTED** [review]
I would not convert back-and-forth to `torch` arrays and instead dispatch the logic depending on the type of `input_ids` and `logits` so as to not add a performance penalty during inference.

File: outlines/processors/base_logits_processor.py (hunk 2 above)
Comment (lapp0) [reply outlines/processors/base_logits_processor.py:66]: The idea is that we can cast `numpy` and `mlx.core` arrays to `torch.Tensor` using shared memory, **not** a copy operation :)

Then we can implement a single `OutlinesLogitsProcessor` class with a single `process_logits()` method and it works out of the box for any library. While not in `main` yet, I've tested `FSMLogitsProcessor`, and `RegexLogitsProcessor` with every nearly `outlines.models` option and they work with no additional changes.

This makes the processors easy to maintain and implement.

File: outlines/processors/base_logits_processor.py (hunk 1 above)
Comment (lapp0) [reply outlines/processors/base_logits_processor.py:17]: My thinking here was if we want a new logits processor which isn't for structured generation, e.g. `RepetitionPenaltyLogitsProcessor`. We can easily implement this processor once and use it in every `outlines.models` without any changes specific to the model.

This provides us the flexiiblity to have the inference library handle the decoder pass and sampling, and `outlines` can handle all logits augmentation.

Another example: `mlxlm` doesn't support `stop_strings` and `stop_strings` is broken in the latest `transformers` version (`4.41.2`). We could implement a `StopStringsLogitsProcessor` and make it available to all inference libraries. Not arguing that this should be a priority, but I am arguing that this provides us flexibility and opportunities.

**Comment by lapp0** [comment]
> I would not convert back-and-forth to `torch` arrays and instead dispatch the logic depending on the type of `input_ids` and `logits` so as to not add a performance penalty during inference.

Here are the benchmark results:

```
[50.00%] ··· Running (bench_processors.LogitsProcessorBenchmark.time_logits_processor--).
[100.00%] ··· bench_processors.LogitsProcessorBenchmark.time_logits_processor                                                                                                                                   ok
[100.00%] ··· ======== ==========
               param1            
              -------- ----------
               torch    150±20μs 
               numpy    161±7μs  
                mlx     195±20μs 
              ======== ==========
```

The performance penalty for casting `mlx.core.array` -> `torch.Tensor` -> `mlx.core.array` is 45 nanoseconds.
